code/datums/elements/immerse.dm

```diff
@@ -0,0 +1,387 @@
+
+	src.icon = icon
+	src.icon_state = icon_state
+	src.color = color
+	src.alpha = alpha
+	src.mask_icon = mask_icon
+
+	/**
+	 * Hello, you may be wondering why we're blending icons and not simply
+	 * overlaying one mutable appearance with the blend multiply on another.
+	 * Well, the latter option doesn't work as neatly when ultimately added
+	 * to an atom with the KEEP_TOGETHER appearance flag, with the mask also
+	 * showing on said atom, while we don't want it to.
+	 */
+	var/icon/immerse_icon = generated_immerse_icons["[icon]-[icon_state]-[mask_icon]"]
+	if(!immerse_icon)
+		immerse_icon = icon(icon, icon_state)
+		var/icon/sub_mask = icon('icons/effects/effects.dmi', mask_icon)
+		immerse_icon.Blend(sub_mask, ICON_MULTIPLY)
+		immerse_icon = fcopy_rsc(immerse_icon)
+		generated_immerse_icons["[icon]-[icon_state]-[mask_icon]"] = immerse_icon
+
+	RegisterSignal(target, SIGNAL_ADDTRAIT(TRAIT_IMMERSE_STOPPED), PROC_REF(stop_immersion))
+	RegisterSignal(target, SIGNAL_REMOVETRAIT(TRAIT_IMMERSE_STOPPED), PROC_REF(start_immersion))
+
+	if(!HAS_TRAIT(target, TRAIT_IMMERSE_STOPPED))
+		start_immersion(target)
+
+/datum/element/immerse/Detach(turf/source)
+	UnregisterSignal(source, list(SIGNAL_ADDTRAIT(TRAIT_IMMERSE_STOPPED), SIGNAL_REMOVETRAIT(TRAIT_IMMERSE_STOPPED)))
+	if(!HAS_TRAIT(source, TRAIT_IMMERSE_STOPPED))
+		stop_immersion(source)
+	return ..()
+
+///Makes the element start affecting the turf and its contents. Called on Attach() or when TRAIT_IMMERSE_STOPPED is removed.
+/datum/element/immerse/proc/start_immersion(turf/source)
+	SIGNAL_HANDLER
+	RegisterSignals(source, list(COMSIG_ATOM_ABSTRACT_ENTERED, COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZED_ON), PROC_REF(on_init_or_entered))
+	RegisterSignal(source, COMSIG_ATOM_ABSTRACT_EXITED, PROC_REF(on_atom_exited))
+	attached_turfs_and_movables += source
+	for(var/atom/movable/movable as anything in source)
+		on_init_or_entered(source, movable)
+
+///Stops the element from affecting on the turf and its contents. Called on Detach() or when TRAIT_IMMERSE_STOPPED is added.
+/datum/element/immerse/proc/stop_immersion(turf/source)
+	SIGNAL_HANDLER
+	UnregisterSignal(source, list(COMSIG_ATOM_ABSTRACT_ENTERED, COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZED_ON, COMSIG_ATOM_ABSTRACT_EXITED))
+	for(var/atom/movable/movable as anything in attached_turfs_and_movables[source])
+		remove_from_element(source, movable)
+	attached_turfs_and_movables -= source
+
+/**
+ * If the movable is within the right layers and planes, not in the list of movable types to ignore,
+ * or already affected by the element for that matter, Signals will be registered and,
+ * unless the movable (or whatever it's buckled to) is flying, it'll appear as if immersed in that water.
+ */
+/datum/element/immerse/proc/on_init_or_entered(turf/source, atom/movable/movable)
+	SIGNAL_HANDLER
+	if(movable.layer >= ABOVE_ALL_MOB_LAYER || !ISINRANGE(movable.plane, MUTATE_PLANE(FLOOR_PLANE, source), MUTATE_PLANE(GAME_PLANE_UPPER_FOV_HIDDEN, source)))
```

Comment:
doing this based off plane is realllllllly sus

Reply:
If you've a better idea that doesn't involve further cherrypicking for movables that should be ignored, I'm all ears.